Bit aliases is not working in capsule

### Steps to Reproduce

1. git clone git@github.com:valor-software/ngx-bootstrap.git
2. npm i
3. bit init
4. 
```
bit add src/*/ -e 'src/chronos,src/component-loader,src/mini-ngrx,src/positioning,src/spec,src/locale,src/common'
bit add src/chronos src/component-loader src/mini-ngrx src/positioning -n internal
```
5. copy this to the config of bit in package.json:
```
"bit": {
    "env": {
      "compiler": "bit.envs/angular@0.1.3"
    },
    "componentsDefaultDirectory": "components/{name}",
    "packageManager": "npm",
    "resolveModules": {
      "aliases": {
        "ngx-bootstrap": "src"
      }
    },
    "overrides": {
      "*": {
        "dependencies": {
          "@angular/forms": "-",
          "@angular/animations": "-"
        },
        "peerDependencies": {
          "@angular/forms": "^8.0.0",
          "@angular/animations": "^8.0.0"
        }
      }
    }
  }
```
6. import the compiler: https://bit.dev/bit/envs/angular
7. bit s (need to be ok)
8. bit build alert

This is the error you will get:

BUILD ERROR
alert/alert.component.ts(11,26): error TS2307: Cannot find module 'ngx-bootstrap/utils'.

### Screenshots, exceptions and logs

If applicable, add screenshots, exceptions, and logs to help explain your problem.
![image](https://user-images.githubusercontent.com/35892475/61875602-ae88d600-aef3-11e9-98dc-8536c1e6f6bb.png)

### Specifications

- Bit version: 14.2.2
- Node version: 12.7.0
- npm  version: 6.10.0
- Platform: macOS
- Bit compiler (include version): https://bit.dev/bit/envs/angular?version=0.1.3

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -7,7 +7,7 @@ and this project adheres to [Semantic Versioning](http://semver.org/).
 
 ## [unreleased]
 
-- [#1856](https://github.com/teambit/bit/issues/1856) expose `writeLinks` function for the compiler to run after installing packages on a capsule
+- [#1856](https://github.com/teambit/bit/issues/1856) fix links deletion from `node_modules` after installing packages by a compiler on a capsule
 - [#1710](https://github.com/teambit/bit/issues/1710) improve performance of importing an entire collection
 
 ## [14.2.2] - 2019-07-24

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -7,6 +7,7 @@ and this project adheres to [Semantic Versioning](http://semver.org/).
 
 ## [unreleased]
 
+- [#1856](https://github.com/teambit/bit/issues/1856) expose `writeLinks` function for the compiler to run after installing packages on a capsule
 - [#1710](https://github.com/teambit/bit/issues/1710) improve performance of importing an entire collection
 
 ## [14.2.2] - 2019-07-24

--- a/src/consumer/component-ops/build-component.js
+++ b/src/consumer/component-ops/build-component.js
@@ -320,7 +320,7 @@ const _runBuild = async ({
           await isolator.installPackagesOnRoot(packages);
           // after installing packages on capsule root, some links/symlinks from node_modules might
           // be deleted. rewrite the links to recreate them.
-          await isolator.writeLinks();
+          await isolator.writeLinksOnNodeModules();
         };
         const capsuleFiles = componentWithDependencies.component.files;
         return {

--- a/src/consumer/component-ops/build-component.js
+++ b/src/consumer/component-ops/build-component.js
@@ -316,6 +316,12 @@ const _runBuild = async ({
           }
           return updatedFiles;
         };
+        const installPackages = async (packages: string[] = []) => {
+          await isolator.installPackagesOnRoot(packages);
+          // after installing packages on capsule root, some links/symlinks from node_modules might
+          // be deleted. rewrite the links to recreate them.
+          await isolator.writeLinks();
+        };
         const capsuleFiles = componentWithDependencies.component.files;
         return {
           capsule: isolator.capsule,
@@ -325,6 +331,7 @@ const _runBuild = async ({
           getDependenciesLinks,
           writeLinks: () => isolator.writeLinks(),
           capsuleExec: (cmd, options) => isolator.capsuleExec(cmd, options),
+          installPackages,
           addSharedDir
         };
       };

--- a/src/consumer/component-ops/build-component.js
+++ b/src/consumer/component-ops/build-component.js
@@ -323,6 +323,8 @@ const _runBuild = async ({
           componentWithDependencies,
           writeDists,
           getDependenciesLinks,
+          writeLinks: () => isolator.writeLinks(),
+          capsuleExec: (cmd, options) => isolator.capsuleExec(cmd, options),
           addSharedDir
         };
       };

--- a/src/consumer/component/sources/data-to-persist.js
+++ b/src/consumer/component/sources/data-to-persist.js
@@ -71,10 +71,6 @@ export default class DataToPersist {
   async persistAllToCapsule(capsule: Capsule) {
     this._log();
     this._validateRelative();
-    // const filesForCapsule = this.files.reduce(
-    //   (acc, current) => Object.assign(acc, { [current.path]: current.contents }),
-    //   {}
-    // );
     await Promise.all(this.remove.map(pathToRemove => capsule.removePath(pathToRemove.path)));
     await Promise.all(this.files.map(file => capsule.outputFile(file.path, file.contents)));
     await Promise.all(this.symlinks.map(symlink => this.atomicSymlink(capsule, symlink)));
@@ -113,7 +109,15 @@ export default class DataToPersist {
    */
   toConsole() {
     console.log(`\nfiles: ${this.files.map(f => f.path).join('\n')}`); // eslint-disable-line no-console
+    console.log(`\nsymlinks: ${this.symlinks.map(s => `src: ${s.src}, dest: ${s.dest}`).join('\n')}`); // eslint-disable-line no-console
     console.log(`remove: ${this.remove.map(r => r.path).join('\n')}`); // eslint-disable-line no-console
+  }
+  filterByPath(filterFunc: Function): DataToPersist {
+    const dataToPersist = new DataToPersist();
+    dataToPersist.addManyFiles(this.files.filter(f => filterFunc(f.path)));
+    dataToPersist.removeManyPaths(this.remove.filter(r => filterFunc(r.path)));
+    dataToPersist.addManySymlinks(this.symlinks.filter(s => filterFunc(s.dest)));
+    return dataToPersist;
   }
   async _persistFilesToFS() {
     return Promise.all(this.files.map(file => file.write()));

--- a/src/environment/isolator.js
+++ b/src/environment/isolator.js
@@ -29,6 +29,7 @@ export default class Isolator {
   capsuleBitMap: BitMap;
   capsulePackageJson: PackageJsonFile; // this is the same packageJson of the main component as it located on the root
   componentWithDependencies: ComponentWithDependencies;
+  manyComponentsWriter: ManyComponentsWriter;
   constructor(capsule: Capsule, scope: Scope, consumer?: ?Consumer) {
     this.capsule = capsule;
     this.scope = scope;
@@ -54,7 +55,6 @@ export default class Isolator {
     }
     const writeToPath = opts.writeToPath;
     const concreteOpts = {
-      // consumer: this.consumer,
       componentsWithDependencies: [componentWithDependencies],
       writeToPath,
       override: opts.override,
@@ -74,24 +74,35 @@ export default class Isolator {
       capsule: this.capsule
     };
     this.componentWithDependencies = componentWithDependencies;
-    // $FlowFixMe
-    const manyComponentsWriter = new ManyComponentsWriter(concreteOpts);
+    this.manyComponentsWriter = new ManyComponentsWriter(concreteOpts);
+    await this.writeComponentsAndDependencies();
+    await this.installPackages();
+    await this.writeLinks();
+    this.capsuleBitMap = this.manyComponentsWriter.bitMap;
+    return componentWithDependencies;
+  }
+
+  async writeComponentsAndDependencies() {
     logger.debug('ManyComponentsWriter, writeAllToIsolatedCapsule');
     this._manipulateDir();
-    await manyComponentsWriter._populateComponentsFilesToWrite();
-    await manyComponentsWriter._populateComponentsDependenciesToWrite();
+    await this.manyComponentsWriter._populateComponentsFilesToWrite();
+    await this.manyComponentsWriter._populateComponentsDependenciesToWrite();
     await this._persistComponentsDataToCapsule();
+  }
+
+  async installPackages() {
     // $FlowFixMe
-    this.capsulePackageJson = componentWithDependencies.component.packageJsonFile;
+    this.capsulePackageJson = this.componentWithDependencies.component.packageJsonFile;
     // $FlowFixMe
-    const componentRootDir: string = componentWithDependencies.component.writtenPath;
+    const componentRootDir: string = this.componentWithDependencies.component.writtenPath;
     await this._addComponentsToRoot(componentRootDir);
     logger.debug('ManyComponentsWriter, install packages on capsule');
     await this._installWithPeerOption(componentRootDir);
-    const links = await manyComponentsWriter._getAllLinks();
+  }
+
+  async writeLinks() {
+    const links = await this.manyComponentsWriter._getAllLinks();
     await links.persistAllToCapsule(this.capsule);
-    this.capsuleBitMap = manyComponentsWriter.bitMap;
-    return componentWithDependencies;
   }
 
   _manipulateDir() {
@@ -162,20 +173,7 @@ export default class Isolator {
   }
 
   async _getNpmVersion() {
-    const execResults = await this.capsule.exec('npm --version');
-    const versionString = await new Promise((resolve, reject) => {
-      let version = '';
-      execResults.stdout.on('data', (data: string) => {
-        version += data;
-      });
-      execResults.stdout.on('error', (error: string) => {
-        return reject(error);
-      });
-      // @ts-ignore
-      execResults.on('close', () => {
-        return resolve(version);
-      });
-    });
+    const versionString = await this.capsuleExec('npm --version');
     const validVersion = semver.coerce(versionString);
     return validVersion ? validVersion.raw : null;
   }
@@ -194,7 +192,11 @@ export default class Isolator {
       );
     }
     const args = ['install', ...modules, '--no-save'];
-    const execResults = await this.capsule.exec(`npm ${args.join(' ')}`, { cwd: directory });
+    return this.capsuleExec(`npm ${args.join(' ')}`, { cwd: directory });
+  }
+
+  async capsuleExec(cmd: string, options?: ?Object): Promise<string> {
+    const execResults = await this.capsule.exec(cmd, options);
     let output = '';
     return new Promise((resolve, reject) => {
       execResults.stdout.on('data', (data: string) => {

--- a/src/environment/isolator.js
+++ b/src/environment/isolator.js
@@ -21,6 +21,7 @@ import { topologicalSortComponentDependencies } from '../scope/graph/components-
 import DataToPersist from '../consumer/component/sources/data-to-persist';
 import BitMap from '../consumer/bit-map';
 import { getManipulateDirForComponentWithDependencies } from '../consumer/component-ops/manipulate-dir';
+import GeneralError from '../error/general-error';
 
 export default class Isolator {
   capsule: Capsule;
@@ -30,6 +31,8 @@ export default class Isolator {
   capsulePackageJson: PackageJsonFile; // this is the same packageJson of the main component as it located on the root
   componentWithDependencies: ComponentWithDependencies;
   manyComponentsWriter: ManyComponentsWriter;
+  _npmVersionHasValidated: boolean = false;
+  componentRootDir: string;
   constructor(capsule: Capsule, scope: Scope, consumer?: ?Consumer) {
     this.capsule = capsule;
     this.scope = scope;
@@ -76,7 +79,7 @@ export default class Isolator {
     this.componentWithDependencies = componentWithDependencies;
     this.manyComponentsWriter = new ManyComponentsWriter(concreteOpts);
     await this.writeComponentsAndDependencies();
-    await this.installPackages();
+    await this.installComponentPackages();
     await this.writeLinks();
     this.capsuleBitMap = this.manyComponentsWriter.bitMap;
     return componentWithDependencies;
@@ -90,14 +93,14 @@ export default class Isolator {
     await this._persistComponentsDataToCapsule();
   }
 
-  async installPackages() {
+  async installComponentPackages() {
     // $FlowFixMe
     this.capsulePackageJson = this.componentWithDependencies.component.packageJsonFile;
     // $FlowFixMe
-    const componentRootDir: string = this.componentWithDependencies.component.writtenPath;
-    await this._addComponentsToRoot(componentRootDir);
+    this.componentRootDir = this.componentWithDependencies.component.writtenPath;
+    await this._addComponentsToRoot();
     logger.debug('ManyComponentsWriter, install packages on capsule');
-    await this._installWithPeerOption(componentRootDir);
+    await this._installWithPeerOption();
   }
 
   async writeLinks() {
@@ -145,13 +148,13 @@ export default class Isolator {
     await dataToPersist.persistAllToCapsule(this.capsule);
   }
 
-  async _addComponentsToRoot(rootDir: string): Promise<void> {
+  async _addComponentsToRoot(): Promise<void> {
     const capsulePath = this.capsule.container.getPath();
     // the capsulePath hack only works for the fs-capsule
     // for other capsule types, we would need to do this
     // (and other things) inside the capsule itself
     // rather than fetching its folder and using it
-    const rootPathInCapsule = path.join(capsulePath, rootDir);
+    const rootPathInCapsule = path.join(capsulePath, this.componentRootDir);
     const componentsToAdd = this.componentWithDependencies.allDependencies.reduce((acc, component) => {
       // $FlowFixMe - writtenPath is defined
       const componentPathInCapsule = path.join(capsulePath, component.writtenPath);
@@ -178,21 +181,27 @@ export default class Isolator {
     return validVersion ? validVersion.raw : null;
   }
 
-  async _installPackages(directory: string, modules: string[] = []) {
+  async installPackagesOnRoot(modules: string[] = []) {
+    await this._throwForOldNpmVersion();
+    const args = ['install', ...modules, '--no-save'];
+    return this.capsuleExec(`npm ${args.join(' ')}`, { cwd: this.componentRootDir });
+  }
+
+  async _throwForOldNpmVersion() {
+    if (this._npmVersionHasValidated) {
+      return;
+    }
     const npmVersion = await this._getNpmVersion();
     if (!npmVersion) {
-      return Promise.reject(new Error('Failed to isolate component: unable to run npm'));
+      throw new Error('Failed to isolate component: unable to run npm');
     }
     if (!semver.satisfies(npmVersion, ACCEPTABLE_NPM_VERSIONS)) {
-      return Promise.reject(
-        new Error(
-          `Failed to isolate component: found an old version of npm (${npmVersion}). ` +
-            `To get rid of this error, please upgrade to npm ${ACCEPTABLE_NPM_VERSIONS}`
-        )
+      throw new GeneralError(
+        `Failed to isolate component: found an old version of npm (${npmVersion}). ` +
+          `To get rid of this error, please upgrade to npm ${ACCEPTABLE_NPM_VERSIONS}`
       );
     }
-    const args = ['install', ...modules, '--no-save'];
-    return this.capsuleExec(`npm ${args.join(' ')}`, { cwd: directory });
+    this._npmVersionHasValidated = true;
   }
 
   async capsuleExec(cmd: string, options?: ?Object): Promise<string> {
@@ -218,26 +227,26 @@ export default class Isolator {
    * again. The reason for adding the missing peer into devDependencies is to not get them deleted
    * once `npm install` is running along the road.
    */
-  async _installWithPeerOption(directory: string, installPeerDependencies: boolean = true) {
-    await this._installPackages(directory);
+  async _installWithPeerOption(installPeerDependencies: boolean = true) {
+    await this.installPackagesOnRoot();
     if (installPeerDependencies) {
-      const peers = await this._getPeerDependencies(directory);
+      const peers = await this._getPeerDependencies();
       if (!R.isEmpty(peers)) {
         this.capsulePackageJson.packageJsonObject.devDependencies = Object.assign(
           this.capsulePackageJson.packageJsonObject.devDependencies || {},
           peers
         );
         await this._writeCapsulePackageJson();
-        await this._installPackages(directory);
+        await this.installPackagesOnRoot();
       }
     }
   }
 
-  async _getPeerDependencies(dir: string): Promise<Object> {
+  async _getPeerDependencies(): Promise<Object> {
     const packageManager = DEFAULT_PACKAGE_MANAGER;
     let npmList;
     try {
-      npmList = await this._getNpmListOutput(dir, packageManager);
+      npmList = await this._getNpmListOutput(packageManager);
     } catch (err) {
       logger.error(err);
       throw new Error(
@@ -247,24 +256,13 @@ export default class Isolator {
     return npmClient.getPeerDepsFromNpmList(npmList, packageManager);
   }
 
-  async _getNpmListOutput(dir: string, packageManager: string) {
+  async _getNpmListOutput(packageManager: string): Promise<string> {
     const args = [packageManager, 'list', '-j'];
-    const execResults = await this.capsule.exec(args.join(' '), { cwd: dir });
-    let output = '';
-    let outputErr = '';
-    return new Promise((resolve, reject) => {
-      execResults.stdout.on('data', (data: string) => {
-        output += data;
-      });
-      execResults.stdout.on('error', (error: string) => {
-        outputErr += error;
-      });
-      // @ts-ignore
-      execResults.on('close', () => {
-        if (output) return resolve(output);
-        if (outputErr && outputErr.startsWith('{')) return resolve(output); // it's probably a valid json with errors, that's fine, parse it.
-        return reject(outputErr);
-      });
-    });
+    try {
+      return await this.capsuleExec(args.join(' '), { cwd: this.componentRootDir });
+    } catch (err) {
+      if (err && err.startsWith('{')) return err; // it's probably a valid json with errors, that's fine, parse it.
+      throw err;
+    }
   }
 }

--- a/src/environment/isolator.js
+++ b/src/environment/isolator.js
@@ -108,6 +108,17 @@ export default class Isolator {
     await links.persistAllToCapsule(this.capsule);
   }
 
+  /**
+   * used by compilers that create capsule.
+   * when installing packages on the capsule, the links generated on node_modules may be deleted
+   */
+  async writeLinksOnNodeModules() {
+    const links = await this.manyComponentsWriter._getAllLinks();
+    const nodeModulesLinks = links.filterByPath(filePath => filePath.startsWith('node_modules'));
+    nodeModulesLinks.toConsole();
+    await nodeModulesLinks.persistAllToCapsule(this.capsule);
+  }
+
   _manipulateDir() {
     const allComponents = [this.componentWithDependencies.component, ...this.componentWithDependencies.allDependencies];
     const manipulateDirData = getManipulateDirForComponentWithDependencies(this.componentWithDependencies);

